### PR TITLE
Y24-190-3: Use SS v2 for TagLayoutTemplates

### PR DESCRIPTION
**Pull request by sdjmchattie**
Closes #1818 

#### Changes proposed in this pull request

- Switch to using SS v2 for TagLayoutTemplate
- There is one remaining instance of v1 used for TagLayoutTemplate which I have been unable to work around yet.  I plan to fix this by also implementing TagLayout in SS v2 and then I can push the creation through the correct endpoint.

#### Instructions for Reviewers

_[All PRs] - Confirm PR template filled_  
_[Feature Branches] - Review code_

### CODE REVIEW

File: app/models/labware_creators/tagged_plate.rb [hunk 1]
```diff
@@ -79,6 +79,10 @@ def convert_tag_plate_to_new_purpose
 
     def create_labware!
       create_plate! do |plate_uuid|
+        # TODO: {Y24-190} Work out a way to call the `create!` method on TagLayoutTemplate model in Sequencescape
+        #       using the V2 API. I think either we need to misuse the PATCH method with some kind of
+        #       attributes telling Sequencescape to run the `create!` method, or we need to create a new
+        #       endpoint associated with a TagLayoutTemplate that will run the `create!` method.
```
Comment (KatyTaylor) [inline app/models/labware_creators/tagged_plate.rb:85]: I thought you were planning to redesign it so that you could call `create!` on `TagLayout` instead, and that pulled info from the `TagLayoutTemplate`?

File: app/sequencescape/sequencescape/api/v2/tag_layout_template.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  end
+
+  # Performs the coercion of this instance so that it behaves appropriately given the direction
+  # and walking algorithm information.
+  def coerce
+    extend("limber/tag_layout_template/in_#{direction.gsub(/\s+/, '_')}s".camelize.constantize)
+    extend("limber/tag_layout_template/walk_#{walking_by.gsub(/\s+/, '_')}".camelize.constantize)
+  rescue NameError => e
+    Rails.logger.warn("Unrecognised layout options: #{e.message}")
+    extend Unsupported
```
Comment (KatyTaylor) [inline app/sequencescape/sequencescape/api/v2/tag_layout_template.rb:19]: So this method and the one below are taken from the existing `app > models > limber > tag_layout_template.rb` file... is that file needed any more?

File: spec/factories/tag_factories.rb [hunk 3]
```diff
@@ -12,13 +12,23 @@ def generate_oligo
...
+    tag_group { create :v2_tag_group, v2_tags: [instance] }
   end
 
-  factory :v2_tag_group, class: Sequencescape::Api::V2::Tag do
+  factory :v2_tag_group, class: Sequencescape::Api::V2::TagGroup do
     skip_create
 
+    transient { v2_tags { [] } }
     sequence(:name) { |index| "TagGroup#{index}" }
+    tags { v2_tags.map { |t| { index: t.map_id, oligo: t.oligo } } }
```
Comment (KatyTaylor) [inline spec/factories/tag_factories.rb:25]: So `v2_tags` get passed into this factory as instances of `Sequencescape::Api::V2::Tag`, but this data structure seems to change them into a simple hash? Is that how it appears in real API results too?

**Review by KatyTaylor — APPROVED** [review]
Looks good I think, just a couple of questions.

File: README.md [hunk 4]
```diff
@@ -200,12 +200,68 @@ There are a few tools available to assist with writing specs:
...
+#### Request stubbing for the Sequencescape v1 API
 
 Request stubs are provided by webmock. Two helper methods will assist with the majority of mocking requests to the api, `stub_api_get` and `stub_api_post`. See `spec/support/api_url_helper.rb` for details.
 
 **Note**: Due to the way the api functions, the factories don't yet support nested associations.
 
+#### Request stubbing for the Sequencescape v2 API
+
+The V2 API uses `JsonApiClient::Resource` sub-classes to represent the records in memory.
+Generally these are quite dynamic so you don't need to explicitly specify every property the API will respond with.
```
[the remark this answers is not in the document]
Comment (sdjmchattie) [reply README.md:212]: Magic.  Basically the resources mostly look like this:

```
# frozen_string_literal: true

class Sequencescape::Api::V2::Project < Sequencescape::Api::V2::Base
end
```

but even so, that's just a template for the client gem to know that it should load a resource from that endpoint and the attributes from the API will be available to you when you use this class.  They are dynamically assigned.

The exception is for relationships.  If the API supports a resource having associated other resources, you do need to specify `has_one`, `has_many` etc in order for it to make those accessible via the client gem.

You can also add custom logic to these, like I did for TagLayoutTemplate, so you can change the named algorithms into implementations of that algorithm on the Limber side.

File: app/models/labware_creators/tagged_plate.rb (hunk 1 above)
Comment (sdjmchattie) [reply app/models/labware_creators/tagged_plate.rb:85]: Yes, but we don't have `TagLayout` in V2 yet.  I plan to do that in a soon pull request.  This comment was written before we came across that solution.  I'm not sure it's worth updating since it's coming out in the next week.

File: app/sequencescape/sequencescape/api/v2/tag_layout_template.rb (hunk 2 above)
Comment (sdjmchattie) [reply app/sequencescape/sequencescape/api/v2/tag_layout_template.rb:19]: I'm leaving removal of V1 until the end.  It's still unclear how tightly tied they are into the system and I do still need it until we remove the TODO items.

File: spec/factories/tag_factories.rb (hunk 3 above)
Comment (sdjmchattie) [reply spec/factories/tag_factories.rb:25]: Yes.  I wouldn't have done it this way as it's not how V1 works, but the V2 endpoint was pre-existing with this and if I change it, who knows what else might break.

One of the biggest (and possibly least obvious) issues with APIs is that they're even harder to change than other aspects of systems.  You're setting up a contract with people outside your team's scope for how they can interface with your application.  Making decisions about API implementations should never be taken lightly.  Deploying a bad decision has to be maintained until the next version of the API.
